sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/AliasAwareOutputExpression.scala

```diff
@@ -70,53 +66,16 @@ trait AliasAwareOutputExpression extends SQLConfHelper {
   protected def normalizeExpression(
       expr: Expression,
       pruneFunc: (Expression, AttributeSet) => Option[Expression]): Seq[Expression] = {
-    val normalizedCandidates = new mutable.HashSet[Expression]()
-    normalizedCandidates.add(expr)
     val outputSet = AttributeSet(outputExpressions.map(_.toAttribute))
-
-    def pruneCandidate(candidate: Expression): Option[Expression] = {
+    expr.multiTransform {
```

Comment:
@cloud-fan, @ulysses-you I've updated this PR. Now it is based on `multiTransform` and contains changes from both this PR and https://github.com/apache/spark/pull/39556.

`normalizeExpression()` becomes as simple as this with `multiTransform`.

Please note that currently `pruneFunc` is used only for "after transformation filtering", but, as `multiTransform` does the mapping in "one run" (unlike the removed code which runs a `transform` for each alias) it is much more efficient than the removed version if we have high number of aliases.

Some early pruning would also be possible using `multiTransform`, I will show you that version a bit later.

Reply:
According to my usage. The `multiTransform` should at least support 3 cases of pruning:
1. the max limit size of returned result
2. eagerly pruning func
    - prune the result whose references is not subset of output
    - prune intermediate result if the alias map does not contain any other sub-expression
    - prune sub-expression, e.g. `PartitionCollection(a, b)` -> `PartitionCollection(a)` if b is not subset of output

If all this requirements can be matched, I think it's good to switch to multi-transofrm.